### PR TITLE
Make loss code more uniform across model types

### PR DESCRIPTION
**Pull request by bpopeters**
This pull request makes a number of small changes intended to make the behavior of the loss code easier to understand and extend.

1. `SparsemaxLoss` now takes a `reduction` parameter similar to that used by `nn.NLLLoss` and `nn.CrossEntropyLoss`.
2. The `CopyGeneratorCriterion` class, which did not subclass `nn.Module`, is replaced by `CopyGeneratorLoss`, which does. The code for computing the loss has been made simpler, although this is still a work in progress. In particular, it would be nice if the `CopyGeneratorLoss` had a similar reduction parameter to other loss modules. I do not like that the loss computation for copy models is divided between two places, the criterion module and the `_compute_loss` method of `CopyGeneratorLossCompute`. For other module types, the loss computation happens entirely within the criterion and that seems like the best pattern to follow.
3. Loss function modules (`nn.NLLLoss`, `LabelSmoothingLoss`, `SparsemaxLoss`, and `CopyGeneratorLoss`) are now the first parameter of `LossCompute`'s `__init__()`. The goal of this is to make it easier to just plug in a new loss function similar to the ones defined in pytorch. This is more intuitive and user-friendly than requiring them to write a subclass of `LossCompute`. I hope that further refactoring will make it possible to remove `CopyGeneratorLossCompute` as well.
4. The unk index is now a parameter of the criterion module, analogous to the treatment of the pad index (aka `ignore_index`).
5. Various comments are updated.

### CODE REVIEW

**Comment by vince62s** [comment]
looks good but do you confirm you tested before /after in the 3 different modes ?
@sebastianGehrmann do you have time for a review?

**Comment by bpopeters** [comment]
I did. Results were similar to what they were on master.

File: onmt/modules/copy_generator.py [hunk 1]
```diff
@@ -91,72 +90,73 @@ def forward(self, hidden, attn, src_map):
         # Original probabilities.
         logits = self.linear(hidden)
         logits[:, self.tgt_dict.stoi[inputters.PAD_WORD]] = -float('inf')
-        prob = self.softmax(logits)
+        prob = F.softmax(logits, 1)
 
         # Probability of copying p(z=1) batch.
-        p_copy = self.sigmoid(self.linear_copy(hidden))
+        p_copy = F.sigmoid(self.linear_copy(hidden))
```
Comment (vince62s) [inline onmt/modules/copy_generator.py]: https://pytorch.org/docs/stable/_modules/torch/nn/functional.html
Not sure this is the best way to change this since it's deprecated.

File: onmt/modules/copy_generator.py [hunk 2]
```diff
@@ -1,16 +1,17 @@
 """ Generator module """
-import torch.nn as nn
 import torch
-import torch.cuda
+import torch.nn as nn
+import torch.nn.functional as F
```
Comment (vince62s) [inline onmt/modules/copy_generator.py]: what is the rationale behind this ? see my comment below.

Comment (bpopeters) [reply onmt/modules/copy_generator.py]: You're correct, it should be `torch.sigmoid` instead of `F.sigmoid`. I prefer the function to the module because using the module requires defining an attribute on the `CopyGenerator` which is not strictly necessary. I prefer to only use `nn.Module` subclasses for things that have parameters. I forgot that `torch.nn.functional` was not the way to do that anymore.

**Comment by vince62s** [comment]
ok, I guess I can merge Ben ?
